pyvista/core/composite.py

```diff
@@ -690,3 +704,253 @@ def copy(self, deep=True):
+        if name is None:
+            return FieldAssociation.NONE, pyvista_ndarray([])
+
+        if not data_assoc:
+            raise KeyError(f'"{name}" is missing from all the blocks of this composite dataset.')
+
+        field_asc = data_assoc[0][0]
+        # set the field association to the preference if at least one occurrence
+        # of it exists
+        if field_asc.name.lower() != preference.lower():
+            for field, _, _ in data_assoc:
+                if field.name.lower() == preference:
+                    field_asc = getattr(FieldAssociation, preference.upper())
+                    break
+
+        # Verify array consistency
+        dims: Set[int] = set()
+        dtypes: Set[np.dtype] = set()
+        for block in self:
+            for field, scalars, _ in data_assoc:
+                # only check for the active field association
+                if field != field_asc:
+                    continue
+                dims.add(scalars.ndim)
+                dtypes.add(scalars.dtype)
+
+        if len(dims) > 1:
+            raise ValueError(f'Inconsistent dimensions {dims} in active scalars.')
+
+        # check complex mismatch
+        is_complex = [np.issubdtype(dtype, np.complexfloating) for dtype in dtypes]
+        if any(is_complex) and not all(is_complex):
+            raise ValueError('Inconsistent complex and real data types in active scalars.')
+
+        return field_asc, scalars
+
+    def as_polydata_blocks(self, copy=True):
+        """Convert all the datasets within this MultiBlock to :class:`pyvista.PolyData`.
+
+        Depending on ``copy``, this may return a new
+        :class:`pyvista.MultiBlock` dataset. Blocks that are already
+        :class:`pyvista.PolyData` will not be copied.
+
+        Parameters
+        ----------
+        copy : bool, optional
+            Option to return a shallow copy of this composite dataset. When
+            ``False``, this dataset will be modified in-place.
+
+        Returns
+        -------
+        pyvista.MultiBlock
+            MultiBlock containing only :class:`pyvista.PolyData` datasets.
+
+        Notes
+        -----
+        Null blocks are converted to empty :class:`pyvista.PolyData`
+        objects. Downstream filters that operate on PolyData cannot accept
+        MultiBlocks with null blocks.
+
+        """
+        # we make a shallow copy here to avoid modifying the original dataset
+        dataset = self
+        if copy:
+            dataset = self.copy(deep=False)
+
+        # always convert to polydata
+        for i, block in enumerate(dataset):
+            if block is not None:
+                if isinstance(block, MultiBlock):
+                    dataset[i] = block.as_polydata_blocks()
+                elif not isinstance(block, pyvista.PolyData):
+                    dataset[i] = block.extract_surface()
+                elif copy:
+                    # create a shallow copy
+                    dataset[i] = block.copy(deep=False)
```

Comment:
Not sure I understand all the logic here.

It seemed to me that `copy=True` on input meant "give me a shallow copy of this MultiBlock". But here we do `block.copy(deep=False)` which copies the PolyData inside (even if this is a shallow copy). This means that `copied_multi[i] is original_multi[i]` doesn't hold, which I'd expect from a shallow copy of `original_multi`. On the other hand the two non-identical meshes are shallow copies, so certain changes in one will show up in the latter...

Not sure how to build a mental model around this, and how `copy` behaves. Plus the docstring says

> Blocks that are already PolyData will not be copied.

But then why do we have a `block.copy()` in here in the first place?

Reply:
Looking at this with fresh eyes, the behavior of copy should be to control if the `PolyData` datasets are copied or not. When converting to polydata, we should always be returning a copy of the multiblock itself so we don't overwrite any of the original datasets, and instead we should focus on if we're copying `PolyData`, because that's the only dataset type that we aren't already forced to copy.

Reply:
Fixed in 75c684fa7b8a89eba93dab55ad9c7a9d61715bd0.

Reply:
Fixed in 75c684fa7b8a89eba93dab55ad9c7a9d61715bd0.